internal/app/machined/pkg/system/services/etcd.go

```diff
@@ -359,3 +317,85 @@ func buildInitialCluster(config runtime.Configurator, name, ip string) (initial
 
 	return initial, nil
 }
+
+func blacklistError(s string) error {
+	return fmt.Errorf("extra etcd arg %q is not allowed", s)
+}
+
+// nolint: gocyclo
+func args(config runtime.Configurator) ([]string, error) {
+	hostname, err := os.Hostname()
+	if err != nil {
+		return nil, err
+	}
+
+	metadata, err := metadata.Open()
+	if err != nil {
+		return nil, err
+	}
+
+	ips, err := net.IPAddrs()
+	if err != nil {
+		return nil, fmt.Errorf("failed to discover IP addresses: %w", err)
+	}
+
+	if len(ips) == 0 {
+		return nil, errors.New("failed to discover local IP")
+	}
+
+	a := argumentor.Args{
+		"name":                  hostname,
+		"data-dir":              constants.EtcdDataPath,
+		"listen-peer-urls":      "https://0.0.0.0:2380",
+		"listen-client-urls":    "https://0.0.0.0:2379",
+		"cert-file":             constants.KubernetesEtcdPeerCert,
+		"key-file":              constants.KubernetesEtcdPeerKey,
+		"trusted-ca-file":       constants.KubernetesEtcdCACert,
+		"peer-client-cert-auth": "true",
+		"peer-cert-file":        constants.KubernetesEtcdPeerCert,
+		"peer-trusted-ca-file":  constants.KubernetesEtcdCACert,
+		"peer-key-file":         constants.KubernetesEtcdPeerKey,
+	}
+
+	extraArgs := argumentor.Args(config.Cluster().Etcd().ExtraArgs())
+
+	for k := range a {
+		if extraArgs.Contains(k) {
+			return nil, blacklistError(k)
+		}
+	}
+
+	if extraArgs.Contains("initial-cluster-state") {
+		return nil, blacklistError("initial-cluster-state")
+	}
+
+	a.Set("initial-cluster-state", "new")
+
+	if !extraArgs.Contains("initial-cluster") {
+		existing := config.Machine().Type() == machine.ControlPlane || metadata.Upgraded
```

Comment:
Does it make more sense to set these by default and then let the merge below overwrite them if present?

Having these blocks of statements is nice for grouping purposes, but feels like a bit more work than is needed.

Reply:
That means we would go through the whole discovery process when we don't need to.